### ISSUE TITLE
Demo of new parameter handling possibility

### ISSUE THREAD

**Issue by kbwestfall** [PYPIT/PYPIT#315]
I've made a new branch `parset_test` and added a demo of some code that I pulled from the MaNGA DAP.  The base class of interest is called ParSet in the file parset.py that I put [here](https://github.com/PYPIT/PYPIT/blob/parset_test/pypit/par/parset.py).  The demo script is [here](https://github.com/PYPIT/PYPIT/blob/parset_test/bin/test_parset.py) to give you a flavor of what a pypit implementation would look like and how the parameter set functions in the code.  Here's the output of the script:

```
$ test_parset.py
spectrograph: shane_kast_blue
  pypitdir: None
     setup: False
     ncpus: 1
  calcheck: False
  preponly: False
   redname: None
 useIDname: False
      load:   settings: None
     spect: None

 directory:     master: ./master
   science: ./science
        qa: qa


Input value invalid: new_spectrograph.
Options are: ['apf_levy', 'keck_deimos', 'keck_hires', 'keck_lris_blue', 'keck_lris_red', 'shane_kast_blue', 'shane_kast_red', 'tng_dolores', 'wht_isis_blue']
Input value incorrect type: 5.
Valid types are: [<class 'bool'>]
Input value incorrect type: 3.2.
Valid types are: [<class 'int'>]
Input value incorrect type: test.
Valid types are: [<class 'pypit.par.parset.ParSet'>, <class 'dict'>]
Input value incorrect type: 5.3.
Valid types are: [<class 'str'>]
Elapsed time: 0.0003049999999999997 seconds
```

We can spend some time sprucing up the `pypit.par.ParSet.__repr__` method and adding a `description` attribute, but this is what the current version can do.

**Comment by rcooke-ast** [comment]
I'm not sure I fully understand where this would be used. Is this to replace the format of the settings.BLAH files, or used in the *.pypit files?

Ahhh... I see the indents aren't quite right for some of the lines, but I appreciate the general idea.

**Comment by kbwestfall** [comment]
Yeah, the indents are messed up for the components of PypitRunPar that are also ParSets themselves.  This is part of the "sprucing up" needed for the __repr__ method.

X and I had talked about something like this as a way of refactoring/simplifying parts of arparse.py, namely argflag and spect.  This doesn't deal with the actual parsing of the settings files into a number of ParSets, but just shows an example of what the 'run' ParSet might look like, how it's instantiated, and how you would interact with it in the code.

**Comment by kbwestfall** [comment]
Okay, I've spruced this up and have a more fully worked example.  It wasn't as straight-forward as I was hoping, and there are some limitations to its flexibility, but I think it's a good step forward.  You can see the full implementation of ParSet [here](https://github.com/PYPIT/PYPIT/blob/parset_test/pypit/par/parset.py), but more important is it's application to the derived ParSets [here](https://github.com/PYPIT/PYPIT/blob/parset_test/pypit/par/derived.py).

If you wanted to, e.g., add a new parameter to RunPar, you would:
 - Include the parameter as a keyword argument in the `__init__` function
 - Specify any or all of the following: its default value, available options, data type, whether or not it's a callable function, and a description, using the appropriate dictionaries as you can see being done for the current set of parameters.

You should only have to do this in the code itself.

I've got a new demo of how this would work within pypit, [here](https://github.com/PYPIT/PYPIT/blob/parset_test/examples/parset_demo.ipynb).  A useful outcome of this is that we wouldn't need to carry around default configuration files anymore  All the defaults and the comments are now defined in the pypit.par.derived module, and we can just have the setup.py file write the default configuration files when the package is installed.  I haven't worked this out yet, but if the ParSet class can write a config file, then it should be straight forward to have it write an rst doc file so that any additional parameters or changes to parameter descriptions are automatically propagated to the documentation.

If this looks good, give me the go-ahead and I'll start the major conversion of the current settings files to config files and build the necessary derived ParSets.  If we want, we could put everything into a single `PypitPar` ParSet that branches down to all the relevant lower-level ParSets, which we can discuss at some point.  For the moment, I can try to make build this so that it has a few follow-on ramifications to how `argflag` and `spect` are used in the code.

**Comment by rcooke-ast** [comment]
Looks good! A few questions:
* How do you propose we get users to specify changes to the parameters in their input `.pypit` file? Will that be the same as currently done?
* I like the idea that the settings will be in a single place in the code (if I understand that part correctly), so we won't be supplying any `settings.blah` files. Will it be easy enough to update the docs every time a new parameter is added? It would be good if this was automatically done during the install of the code, so that we don't have to remember to update the docs every time we add or remove some functionality. I think this is a key part of the usability of the code, since it needs to be clear to the user what keyword arguments they can give, and what the options are.
* I think it's a good idea to save the `.cfg` files every time pypit is run, so that the user knows exactly what parameters were used to reduce their data.
* Is the plan to have a single configuration file, or to have separate configuration files for spectrograph settings and reduction settings?
* How do you view spectrograph settings be set? Will we be writing a `keck_deimos.py` file (for example) for every new spectrograph, that would contain all of the config settings? Whatever we choose, we must make this as stripped as possible. Perhaps there can be a base spectrograph class with all of the default settings, and a subclass for every instrument that inherits the properties of the parent class, and only minor tweaks are needed. That way, we could have a single `instruments.py` file that contains the base classes, and the subclasses. Any thoughts?

In any case, this direction does seem like the best way forward.

**Comment by kbwestfall** [comment]
In turn:
 - I think in the short term, we could have a backwards-compatible parser that can read a `.pypit` file, but use it in the new framework.  That shouldn't be difficult to write.  But in the longer-term, we should think about if we want to keep the `.pypit` format or phase it out.  Just speaking for myself, I find the format of a config file just easier to quickly parse by eye, but I fully appreciate the longer history of `.pypit` files and why we might not want to completely move away from them.
 - I think the answer is yes, but I haven't worked out the details yet.
 - Agreed!
 - Good question.  I like the idea of having both parameter sets be consolidated into a single, global parameter set.  It means you're only passing around one group of settings, not two.  I worked through an example of this and it meant that I needed to rename some of the spectrograph and/or reduction parameter names so that the names were unique across both groups.  Having said that, I don't have strong feeling one way or the other ... yet 😀 
 - There are some further complications here, but, like you say, I think we want a base spectrograph class that is abstracted to the point where it can accommodate any of the spectrographs currently served.  I agree that we could then have a single `instruments.py` file that provides the specific instantiation of each instrument case so that these can be kept and maintained in a single place.  In the same way, we need a, say, `pipelines.py` file that will specify the required reduction parameters for the different pypit pipelines (ARMED, ARMLSD, ...).  Then, in a similar way to allowing the user to modify the reduction parameters, we could allow for user-level spectrographs, as long as they can be defined within the framework of the existing spectrograph base parameters.  This should hopefully make it easier to add and test application of pypit to new instruments.
